### DISCUSSION TITLE
Optionals not recognized after extracting, and also not working as extracting parameter

### DISCUSSION THREAD

**Discussion by Philzen** [assertj/assertj#3213] [Q&A]
I am experimenting with https://github.com/Hakky54/log-captor for a migration. Now i'm stumbling over the extraction of Optionals.

The method [LogEvent::getThrowable](https://github.com/Hakky54/log-captor/blob/master/src/main/java/nl/altindag/log/model/LogEvent.java#L89) returns an `Optional<Throwable>` which i would expect to be able to either extract during the assertJ chain like in example (4) below, or even better which would be resolved on-the-fly as in example (3), although i understand that may be a breaking change for some people. Maybe there's a smarter way auto-get the content of the Optional which i haven't discovered in the AssertJ docs yet, always grateful for advice.

The gist is, i'd really like to be able to use `.first()` in the fluent assertion, but the assertion chain seems to break as soon as i try that.

```java
        // Works
        assertThat(logCaptor.getLogEvents().get(0).getThrowable()).isPresent().get(THROWABLE)
            .isExactlyInstanceOf(NullPointerException.class).hasNoCause();

        // Works
        assertThat(logCaptor.getLogEvents().get(0).getThrowable().isPresent()).isTrue();
        assertThat(logCaptor.getLogEvents().get(0).getThrowable().get())
            .isExactlyInstanceOf(NullPointerException.class).hasNoCause();

        // Fails: Instance of Optional instead of NullPointerException (kind of makes sense)
        assertThat(logCaptor.getLogEvents())
            .first().extracting(LogEvent::getThrowable)
            .isExactlyInstanceOf(NullPointerException.class);

        // Fails: Still instance of Optional instead of NullPointerException (!?)
        assertThat(logCaptor.getLogEvents())
            .first().extracting(LogEvent::getThrowable).extracting(Optional::get)
            .isExactlyInstanceOf(NullPointerException.class);

        // not possible (?!)
        assertThat(logCaptor.getLogEvents()).first().extracting(LogEvent::getThrowable)
            .isPresent(); // would like to continue here like in the first working example, but won't compile
```

* assertj core version: 3.24.2
* java version: 11.0.20.1-tem
* test framework version: JUnit 5.8.2

**Comment by joel-costigliola** [comment]
Thanks for reporting this @Philzen, could you provide a minimal reproducible test so that we can try on our side? 
I'm especially in case 4 which should be an instance of a `Throwable`.

Have you tried using `extracting` with a lambda that is not a method reference, so instead of 
```java
extracting(LogEvent::getThrowable).extracting(Optional::get)
```
you would use:
```java
extracting(logEvent -> logEvent .getThrowable().get())
```

**Comment by scordio** [comment]
If I understood your requirement correctly, you could try the following:

```java
assertThat(logCaptor.getLogEvents())
  .first(InstanceOfAssertFactories.type(LogEvent.class))
  .extracting(LogEvent::getThrowable, InstanceOfAssertFactories.OPTIONAL)
  .get(InstanceOfAssertFactories.THROWABLE) // isPresent() called implicitly
  .isExactlyInstanceOf(NullPointerException.class)
  .hasNoCause();
```

or less verbose, if giving up a bit of type-safety is an option:

```java
assertThat(logEvents)
  .first()
  .extracting("throwable.value", THROWABLE)
  .isExactlyInstanceOf(NullPointerException.class)
  .hasNoCause();
```

(static imports omitted for clarity)
